runtime/serde/serde-cbor/common/src/aws/smithy/kotlin/runtime/serde/cbor/encoding/Tag.kt

```diff
@@ -0,0 +1,211 @@
+            UInt(exponentValue.toULong())
+        }
+
+        val mantissaStr = str.replace(".", "")
+        // Check if the mantissa can be represented as a UInt without overflowing.
+        // If not, it will be encoded as a Bignum.
+        val mantissa: Value = try {
+            if (mantissaStr.startsWith("-")) {
+                NegInt(mantissaStr.toLong().absoluteValue.toULong())
+            } else {
+                UInt(mantissaStr.toULong())
+            }
+        } catch (e: NumberFormatException) {
+            val bigMantissa = BigInteger(mantissaStr)
+            if (mantissaStr.startsWith("-")) {
+                NegBigNum(bigMantissa)
+            } else {
+                BigNum(bigMantissa)
+            }
+        }
+
+        Tag(TagId.DECIMAL_FRACTION.value, List(listOf(exponent, mantissa))).encode(into)
+    }
+
+    internal companion object {
+        internal fun decode(buffer: SdkBufferedSource): DecimalFraction {
+            val list = List.decode(buffer).value
+            check(list.size == 2) { "Expected array of length 2 for decimal fraction, got ${list.size}" }
+
+            val (exponent, mantissa) = list
+
+            val sb = StringBuilder()
+
+            // Append mantissa
+            sb.append(
+                when (mantissa) {
+                    is UInt -> mantissa.value.toString()
+                    is NegInt -> "-${mantissa.value}"
+                    is Tag -> when (mantissa.value) {
+                        is NegBigNum -> mantissa.value.value.toString()
+                        is BigNum -> mantissa.value.value.toString()
+                        else -> throw DeserializationException("Expected BigNum or NegBigNum for CBOR tagged decimal fraction mantissa, got ${mantissa.id}")
+                    }
+                    else -> throw DeserializationException("Expected UInt, NegInt, or Tag for CBOR decimal fraction mantissa, got $mantissa")
+                },
+            )
+
+            when (exponent) {
+                is UInt -> { // Positive exponent, suffix with zeroes
+                    sb.append("0".repeat(exponent.value.toInt()))
+                    sb.append(".")
+                }
+                is NegInt -> { // Negative exponent, prefix with zeroes if necessary
+                    val exponentValue = exponent.value.toInt().absoluteValue
+                    val insertIndex = if (sb[0] == '-') 1 else 0
+                    if (exponentValue > sb.length - insertIndex) {
+                        sb.insert(insertIndex, "0".repeat(exponentValue - sb.length + insertIndex))
+                        sb.insert(insertIndex, '.')
+                    } else {
+                        sb.insert(sb.length - exponentValue, '.')
+                    }
+                }
+                else -> throw DeserializationException("Expected integer for CBOR decimal fraction exponent value, got $exponent.")
+            }
+
+            return DecimalFraction(BigDecimal(sb.toString()))
```

Comment:
**Suggestion**: Apologies but I missed this the first time around when commenting on string manipulation for `BigInteger` but same suggestion applies: I suggest we delegate these down to JVM's `BigDecimal`, which has a constructor for accepting a `BigInteger` mantissa (which JVM calls the `unscaledVal`) and an `Int` exponent (which JVM calls the `scale`).

Reply:
Actually I think you did mention doing this for `BigDecimal` and _I_ missed it, either way, I think it's the right change so I'll do it